### DISCUSSION TITLE
How do I interpret this weird font information?

### DISCUSSION THREAD

**Discussion by ErikDz** [pymupdf/PyMuPDF#863] [Q&A] [answered]
Greetings,
When printing the font for a span in a textblock line I get results such as the following:

> TTE2214118o00
> TT6B43Fo00
> TT6B42Eo00
> TT6B42Co00
> TT6B42Co00
> TT6B42Eo00
> TT6B42Co00
> TT6B4D1o00
> TT6B42Do00
> TT6B42Do00
> TT6B42Do00
> TT6B42Do00
> TT6B42Do00
> TT6B42Do00
> TT6B42Do00
> TT6B42Co00
> TT6B42Co00
> TT6B42Eo00
> TT6B42Do00
> TT6B42Co00
> TT6B42Co00
> TT6B42Eo00
> TT6B42Co00
> TT6B42Co00
> TT6B42Eo00
> TT6B42Co00
> TT6B42Co00
> TT6B42Eo00
> TT6B42Co00
> TT6B42Co00
> TT6B42Eo00
> TT6B42Do00
> TT6B4D4o00
> TT6B42Do00
> TT6B4D4o00
> TT6B42Do00
> TT6B42Do00
> TT6B4D4o00
> TT6B42Do00
> TT6B4D4o00
> TT6B42Do00
> TT6B42Do00
> TT6B4D4o00
> TT6B42Do00
> TT6B4D4o00
> TT6B42Do00
> TT6B42Co00
> TT6B42Co00
> TT6B42Eo00
> TT6B4D4o00
> TT6B4D4o00
> TT6B4D4o00
> TT6B4D4o00
> TT6B42Co00
> TT6B42Co00
> TT6B42Eo00
> TT6B4D4o00
> TT6B4D4o00
> TT6B4D4o00
> TT6B42Do00
> TT6B42Do00
> TT6B42Do00

I suppose this aren't completely random and actually aport some meaning due to the constant format. So, what is it?
Many thanks!

**Comment by JorjMcKie** [comment, marked as answer]
Check with `doc.get_page_fonts(page.number)`. For each font you should find an item therein.

**Reply by ErikDz** [reply]
Greetings,
Thanks for answering.
It displays the following:

> AttributeError: 'Document' object has no attribute 'get_page_fonts'

**Reply by JorjMcKie** [reply]
then try the old method name `doc.getPageFontList(page.number)`

**Reply by ErikDz** [reply]
Hi! I get this:

> [(47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (49, 'cff', 'Type1', 'BCAFHM+TT6B43Ao00', 'F3', 'WinAnsiEncoding'), (50, 'cff', 'Type1', 'BCAFHN+TT6B43Bo00', 'F4', 'WinAnsiEncoding'), (51, 'cff', 'Type1', 'BCAFHO+TTE2214118o00', 'F5', 'WinAnsiEncoding'), (52, 'cff', 'Type1', 'BCAFHP+TT6B43Fo00', 'F6', 'WinAnsiEncoding'), (53, 'cff', 'Type1', 'BCAFIA+TT6B440o00', 'F7', 'WinAnsiEncoding'), (54, 'cff', 'Type1', 'BCAFKA+TT6B441o00', 'F8', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (59, 'cff', 'Type1', 'BCAFLC+TT6B4D1o00', 'F10', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (94, 'cff', 'Type1', 'BCAFPC+TT6B4D5o00', 'F12', 'WinAnsiEncoding'), (95, 'cff', 'Type1', 'BCAFPD+TT6B4D6o00', 'F13', 'WinAnsiEncoding'), (96, 'cff', 'Type1', 'BCAFPE+TT6B4D7o00', 'F14', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (94, 'cff', 'Type1', 'BCAFPC+TT6B4D5o00', 'F12', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 
> 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (96, 'cff', 'Type1', 'BCAFPE+TT6B4D7o00', 'F14', 'WinAnsiEncoding'), (97, 'cff', 'Type1', 'BCAGAF+TT6B4D8o00', 'F15', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (49, 'cff', 'Type1', 'BCAFHM+TT6B43Ao00', 'F3', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (96, 'cff', 'Type1', 'BCAFPE+TT6B4D7o00', 'F14', 'WinAnsiEncoding'), (97, 'cff', 'Type1', 'BCAGAF+TT6B4D8o00', 'F15', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (96, 'cff', 'Type1', 'BCAFPE+TT6B4D7o00', 'F14', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (49, 'cff', 'Type1', 'BCAFHM+TT6B43Ao00', 'F3', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (98, 'cff', 'Type1', 'BCAGEE+TT6B4D9o00', 'F16', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (49, 'cff', 'Type1', 'BCAFHM+TT6B43Ao00', 'F3', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (107, 'cff', 'Type1', 'BCAGGE+TT6B4DAo00', 'F17', ''), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (49, 'cff', 'Type1', 'BCAFHM+TT6B43Ao00', 'F3', 'WinAnsiEncoding'), (50, 'cff', 'Type1', 'BCAFHN+TT6B43Bo00', 'F4', 'WinAnsiEncoding'), (51, 'cff', 'Type1', 'BCAFHO+TTE2214118o00', 'F5', 'WinAnsiEncoding'), (52, 'cff', 'Type1', 'BCAFHP+TT6B43Fo00', 'F6', 'WinAnsiEncoding'), (53, 'cff', 'Type1', 'BCAFIA+TT6B440o00', 'F7', 'WinAnsiEncoding'), (54, 'cff', 'Type1', 'BCAFKA+TT6B441o00', 'F8', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (59, 'cff', 'Type1', 'BCAFLC+TT6B4D1o00', 'F10', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (94, 'cff', 'Type1', 'BCAFPC+TT6B4D5o00', 'F12', 'WinAnsiEncoding'), (95, 'cff', 'Type1', 'BCAFPD+TT6B4D6o00', 'F13', 'WinAnsiEncoding'), (96, 'cff', 'Type1', 'BCAFPE+TT6B4D7o00', 'F14', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (94, 'cff', 'Type1', 'BCAFPC+TT6B4D5o00', 'F12', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (96, 'cff', 'Type1', 'BCAFPE+TT6B4D7o00', 'F14', 'WinAnsiEncoding'), (97, 'cff', 'Type1', 'BCAGAF+TT6B4D8o00', 'F15', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), 
> (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (49, 'cff', 'Type1', 'BCAFHM+TT6B43Ao00', 'F3', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 
> 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (96, 'cff', 'Type1', 'BCAFPE+TT6B4D7o00', 'F14', 'WinAnsiEncoding'), (97, 'cff', 'Type1', 'BCAGAF+TT6B4D8o00', 'F15', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (96, 'cff', 'Type1', 'BCAFPE+TT6B4D7o00', 'F14', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (49, 'cff', 'Type1', 'BCAFHM+TT6B43Ao00', 'F3', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (98, 'cff', 'Type1', 'BCAGEE+TT6B4D9o00', 'F16', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (49, 'cff', 'Type1', 'BCAFHM+TT6B43Ao00', 'F3', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (107, 'cff', 'Type1', 'BCAGGE+TT6B4DAo00', 'F17', ''), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (49, 'cff', 'Type1', 'BCAFHM+TT6B43Ao00', 'F3', 'WinAnsiEncoding'), (50, 'cff', 'Type1', 'BCAFHN+TT6B43Bo00', 'F4', 'WinAnsiEncoding'), (51, 'cff', 'Type1', 'BCAFHO+TTE2214118o00', 'F5', 'WinAnsiEncoding'), (52, 'cff', 'Type1', 'BCAFHP+TT6B43Fo00', 'F6', 'WinAnsiEncoding'), (53, 'cff', 'Type1', 'BCAFIA+TT6B440o00', 'F7', 'WinAnsiEncoding'), (54, 'cff', 'Type1', 'BCAFKA+TT6B441o00', 'F8', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (59, 'cff', 'Type1', 'BCAFLC+TT6B4D1o00', 'F10', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 
> 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), 
> (94, 'cff', 'Type1', 'BCAFPC+TT6B4D5o00', 'F12', 'WinAnsiEncoding'), (95, 'cff', 'Type1', 'BCAFPD+TT6B4D6o00', 'F13', 'WinAnsiEncoding'), (96, 'cff', 'Type1', 'BCAFPE+TT6B4D7o00', 'F14', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding'), (94, 'cff', 'Type1', 'BCAFPC+TT6B4D5o00', 'F12', 'WinAnsiEncoding'), (47, 'cff', 'Type1', 'BCAFEL+TT6B42Eo00', 'F1', 'WinAnsiEncoding'), (48, 'cff', 'Type1', 'BCAFGL+TT6B42Co00', 'F2', 'WinAnsiEncoding'), (58, 'cff', 'Type1', 'BCAFLB+TT6B42Do00', 'F9', 'WinAnsiEncoding'), (93, 'cff', 'Type1', 'BCAFPB+TT6B4D4o00', 'F11', 'WinAnsiEncoding')

How can I get what's in bold text from this?

**Reply by JorjMcKie** [reply]
A particularly crazy page, I must say!
You could load font by font out of that list and look at its information. Here the example from the PyMuPDF page 0:
```python
>>> for item in doc.getPageFontList(page.number):
	f = doc.extractFont(item[0])
	if f[-1] : # fontbuffer exists?
		font = fitz.Font(fontbuffer=f[-1])
		print("Font: %s\n" % font.name, font.flags)

		
Font: Nimbus Sans L Bold
 {'mono': 0, 'serif': 1, 'bold': 1, 'italic': 0, 'substitute': 0, 'stretch': 0, 'fake-bold': 0, 'fake-italic': 0, 'opentype': 0, 'invalid-bbox': 0}
Font: Nimbus Sans L Bold Italic
 {'mono': 0, 'serif': 1, 'bold': 1, 'italic': 1, 'substitute': 0, 'stretch': 0, 'fake-bold': 0, 'fake-italic': 0, 'opentype': 0, 'invalid-bbox': 0}
>>> 
```

**Reply by ErikDz** [reply]
Wow! I didn't know about the extractFont function!
Awesome! Again, thank you very much!
I'm learning a lot thanks to you! :D

**Reply by ErikDz** [reply]
Hey, it didn't work on the PDF I was working on. I don't know why.

This is a sample output:

> Font: TT6B4DAo00 Regular
>  {'mono': 0, 'serif': 1, 'bold': 0, 'italic': 0, 'substitute': 0, 'stretch': 0, 'fake-bold': 0, 'fake-italic': 0, 'opentype': 0, 'invalid-bbox': 0}

It's thousands like that.
I've tried the following: 
```
if font.flags['bold'] == 0:
    print("This is bold")
```
But no font seems to be bold...
[This is the problematic PDF](https://github.com/pymupdf/PyMuPDF/files/5862753/9608_w15_ms_41.pdf)

**Reply by JorjMcKie** [reply]
> But no font seems to be bold...

This value of the dictionary is the result of "asking" the font for the property. If ``font.flags["bold"] != 1``, then the font thinks, it is not bold.
It may be that
1. the font is sloppily created and does not correctly respond.
2. the font indeed is **_not bold,_** but PDF uses "double printing" to simulate bold

In any case, there is nothing you can do ☹

**Reply by ErikDz** [reply]
Ohhh mannn. How would you personally separate the answers for this pdf then?
Many thanks

**Reply by ErikDz** [reply]
I think I found a way!
Please correct it if it's wrong:

> if  s["bbox"][0] < 50 and s["text"][0].isdecimal():

I think it works!
:D

**Reply by JorjMcKie** [reply]
Exactly! I was about to submit the same - but my Internet just broke down.
Unfortunately, the quationnaire maker has changed how he does that.
So flexibility on your side is required ...
